super.c

```diff
@@ -196,6 +244,253 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
     return 0;
 }
 
+/* journal relation code */
+
+static journal_t *simplefs_get_dev_journal(struct super_block *sb,
+                                           dev_t journal_dev)
+{
+    struct simplefs_sb_info *sbi = SIMPLEFS_SB(sb);
+    struct buffer_head *bh;
+    struct block_device *bdev;
+    int hblock, blocksize;
+    unsigned long long sb_block, start, len;
+    unsigned long offset;
+    journal_t *journal;
+    int errno = 0;
+
+#if SIMPLEFS_AT_LEAST(6, 8, 0)
+    struct bdev_handle *bdev_handle;
+    bdev_handle = bdev_open_by_dev(
+        journal_dev, BLK_OPEN_READ | BLK_OPEN_WRITE | BLK_OPEN_RESTRICT_WRITES,
+        sb, &fs_holder_ops);
+#elif SIMPLEFS_AT_LEAST(6, 7, 0)
+    struct bdev_handle *bdev_handle;
+    up_write(&sb->s_umount);
+    bdev_handle = bdev_open_by_dev(journal_dev, BLK_OPEN_READ | BLK_OPEN_WRITE,
+                                   sb, &fs_holder_ops);
+    down_write(&sb->s_umount);
+#elif SIMPLEFS_AT_LEAST(6, 6, 0)
+    up_write(&sb->s_umount);
+    bdev = blkdev_get_by_dev(journal_dev, BLK_OPEN_READ | BLK_OPEN_WRITE, sb,
+                             &fs_holder_ops);
+    down_write(&sb->s_umount);
+#elif SIMPLEFS_AT_LEAST(6, 5, 0)
+    bdev = blkdev_get_by_dev(journal_dev, BLK_OPEN_READ | BLK_OPEN_WRITE, sb,
+                             NULL);
+#elif SIMPLEFS_AT_LEAST(5, 10, 0)
+    bdev = blkdev_get_by_dev(journal_dev, FMODE_READ | FMODE_WRITE | FMODE_EXCL,
+                             sb);
+#endif
+
+
+#if SIMPLEFS_AT_LEAST(6, 8, 0)
+    if (IS_ERR(bdev_handle)) {
+        printk(KERN_ERR
+               "failed to open journal device unknown-block(%u,%u) %ld\n",
+               MAJOR(journal_dev), MINOR(journal_dev), PTR_ERR(bdev_handle));
+        return ERR_CAST(bdev_handle);
+    }
+    bdev = bdev_handle->bdev;
+#elif SIMPLEFS_AT_LEAST(5, 10, 0)
+    if (IS_ERR(bdev)) {
+        printk(KERN_ERR "failed to open block device (%u:%u), error: %ld\n",
+               MAJOR(journal_dev), MINOR(journal_dev), PTR_ERR(bdev));
+        return ERR_CAST(bdev);
+    }
+#endif
+
+    blocksize = sb->s_blocksize;
+    hblock = bdev_logical_block_size(bdev);
+
+    if (blocksize < hblock) {
+        pr_err("blocksize too small for journal device\n");
+        errno = -EINVAL;
+        goto out_bdev;
+    }
+
+    sb_block = SIMPLEFS_BLOCK_SIZE / blocksize;
+    offset = SIMPLEFS_BLOCK_SIZE % blocksize;
+    set_blocksize(bdev, blocksize);
+    bh = __bread(bdev, sb_block, blocksize);
+
+    if (!bh) {
+        pr_err("couldn't read superblock of external journal\n");
+        errno = -EINVAL;
+        goto out_bdev;
+    }
+
+    /*
+     * FIXME:
```

Comment:
Improve this comment to address the problem you are facing and place the possible solution for future work.